### DISCUSSION TITLE
Question about Circuit Breaking Logic in linkerd-proxy: Attempts to Send Requests to Circuit-Broken Targets

### DISCUSSION THREAD

**Discussion by YHM404** [linkerd/linkerd2#12350] [Q&A] [answered]
I've been reading the code for linkerd-proxy and noticed that it first selects a target using EWMA (Exponentially Weighted Moving Average), and then applies circuit breaker logic to this target. If the target is in a "closed" state, the request is rejected. This means that even if a target is in a circuit-broken state, attempts are still made to send requests to it, instead of selecting another target that hasn't been circuit-broken.

Or, could you provide the logic for filtering out routes that have already been circuit-broken?

**Comment by olix0r** [comment, marked as answer]
You're missing an important aspect of the Service contract: [readiness](https://docs.rs/tower-service/latest/tower_service/trait.Service.html#tymethod.poll_ready).

The P2C balancer uses a [ReadyCache](https://docs.rs/tower/latest/tower/ready_cache/cache/struct.ReadyCache.html) to hold its inner endpoint services. The ReadyCache is responsible for only exposing an endpoint to the balancer when its Service::poll_ready returns Poll::Ready(Ok(())).

The circuit breaking behavior occurs within an individual endpoint service. When the circuit breaker is tripped, the endpoint's `Service::poll_ready` returns `Poll::Pending` until requests are allowed again, and so the balancer simply does not consider it for requests until the readiness changes.

**Reply by YHM404** [reply]
I see :)
Thanks for your reply!
